### PR TITLE
feat: Add a description to a grist table column 

### PR DESCRIPTION
**Pull request by CamilleLegeron**
Issue: #368 

Description in right panel :
image

Info in Detail View (only if there is a description)
image

### CODE REVIEW

**Comment by berhalak** [comment]
Thanks for the preview. Some initial thoughts/suggestions about the UX on the GridView:
- The icon should only be shown when the description is set.
- Icon should be left aligned, but it shouldn't disrupt the label's position (text should still be in the center)
- When a user double-clicks the column header the popup should be shown to rename and describe the column. Now, it is shown when we click the `(i)` icon. So we are removing the old way of renaming columns.
- Tooltip with the description should be shown only when we hover over the icon. (So there are two places where user can see the description: either by double-clicking the column header or by hovering over the `(i)` icon).

**Comment by CamilleLegeron** [comment]
> Thanks for the preview. Some initial thoughts/suggestions about the UX on the GridView:
> 
> * The icon should only be shown when the description is set.
> * When a user double-clicks the column header the popup should be shown to rename and describe the column. Now, it is shown when we click the `(i)` icon. So we are removing the old way of renaming columns.
> * Tooltip with the description should be shown only when we hover over the icon. (So there are two places where the user can see the description: by double-clicking the column header or hovering over the `(i)` icon).

Thank you @berhalak, after conversation on the issue #368 we divided the work into two branches and remove from that one the work on GridView, but the behavior you describe seems perfect to me (@yohanboniface)

File: app/client/ui/FieldConfig.ts [hunk 1]
```diff
@@ -88,6 +89,42 @@ export function buildNameConfig(
...
+    cursor.subscribe(() => {
+      editor?.blur();
+    })
+  );
+
+  return [
+    cssLabel(t("DESCRIPTION")),
+    cssRow(
+      editor = cssTextArea(fromKo(origColumn.description),
+        { onInput: false },
```
Comment (berhalak) [inline app/client/ui/FieldConfig.ts:115]:
```suggestion
        { onInput: false },
          { rows: '3' },
```

File: app/client/ui/FieldConfig.ts [hunk 2]
```diff
@@ -88,6 +89,42 @@ export function buildNameConfig(
...
+        { placeholder: t("If necesary, describe the column") },
+        dom.on('input', async (e, elem) => {
```
Comment (berhalak) [inline app/client/ui/FieldConfig.ts:117]:
```suggestion
        dom.on('blur', async (e, elem) => {
```

File: app/client/ui/tooltips.ts [hunk 3]
```diff
@@ -240,43 +240,59 @@ export function tooltipCloseButton(ctl: ITooltipControl): HTMLElement {
 }
 
 /**
- * Renders an info icon that shows a tooltip with the specified `content` on click.
+ * Renders an icon that shows a tooltip with the specified `content` on click.
  */
+function iconTooltip(
```
Comment (berhalak) [inline app/client/ui/tooltips.ts:245]: I would leave the things here as they were and would just create the iconTooltip helper for the card view.

**Comment by berhalak** [comment]
@CamilleLegeron, Can you add some tests for this field in the `test/nbrowser` folder?

File: app/client/ui/tooltips.ts (hunk 3 above)
Comment (CamilleLegeron) [reply app/client/ui/tooltips.ts:245]: I am creating a helper for the card view but it is a lot of code duplicated for only one icon difference and one style (`display: inline-block`)

File: app/client/ui/FieldConfig.ts [hunk 4]
```diff
@@ -88,6 +89,42 @@ export function buildNameConfig(
   ];
 }
 
+export function buildDescriptionConfig(
+  owner: MultiHolder,
+  origColumn: ColumnRec,
+  cursor: ko.Computed<CursorPos>,
+) {
+  const editedDescription = Observable.create(owner, '');
```
[the remark this answers is not in the document]
Comment (CamilleLegeron) [reply app/client/ui/FieldConfig.ts:97]: If we do not use an observable, what type of object should we use ?

[the remark this answers is not in the document]
Comment (berhalak) [reply app/client/ui/FieldConfig.ts:97]: But I don't think it is used anywhere here, is it?. Textarea is bound to the origColumn.description, and the `editedDescirption` is actually never used in the UI.

**Comment by CamilleLegeron** [comment]
> @CamilleLegeron, Can you add some tests for this field in the `test/nbrowser` folder?

Hi ! I just add simple tests, we can perhaps complete them with the next step (make this feature appear in the grid view). What do you think ?

**Comment by berhalak** [comment]
> > @CamilleLegeron, Can you add some tests for this field in the `test/nbrowser` folder?
> 
> Hi ! I just add simple tests, we can perhaps complete them with the next step (make this feature appear in the grid view). What do you think ?

I would rather prefer to have these tests in this PR. The one with GridView will be much bigger, as the columns header will be rewritten.

**Comment by CamilleLegeron** [comment]
> > > @CamilleLegeron, Can you add some tests for this field in the `test/nbrowser` folder?
> > 
> > 
> > Hi ! I just add simple tests, we can perhaps complete them with the next step (make this feature appear in the grid view). What do you think ?
> 
> I would rather prefer to have these tests in this PR. The one with GridView will be much bigger, as the columns header will be rewritten.

In this PR there is no change of GridView, only DetailView is affected, I can add a test for that but I don't know what other relevant test we could add.

**Review by berhalak — CHANGES_REQUESTED** [review]
Hi @CamilleLegeron,

All looks good except the `editedDescription` variable in `FieldConfig`, and the styling of the Description field. 

I suggested removing the outline from the textbox (when there the textarea has a focus, there is a black border), and adding some padding. I think that moving styles that are now in the `&[readonly]` section, to the main definition should be enough.

One more thing, which I haven't realized earlier. We have a text area, so we can enter a multiline description, but the tooltip shows everything in one line.

File: app/client/ui/FieldConfig.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (CamilleLegeron) [reply app/client/ui/FieldConfig.ts:97]: In fact I'd forgotten this point and your right it's useless

**Comment by CamilleLegeron** [comment]
> Hi @CamilleLegeron,
> 
> All looks good except the `editedDescription` variable in `FieldConfig`, and the styling of the Description field.
> 
> I suggested removing the outline from the textbox (when the textarea has a focus, there is a black border), and adding some padding. I think that moving styles that are now in the `&[readonly]` section, to the main definition should be enough.
> 
> One more thing, which I haven't realized earlier. We have a text area, so we can enter a multiline description, but the tooltip shows everything in one line.

Hi @berhalak, I have done what you asked in my last commit

**Comment by anaisconce** [comment]
Thank you, @CamilleLegeron and @berhalak !

@berhalak pushed the preview to https://grist-column-labels.fly.dev/ so I could play with it. Sorry I didn't get my comments in before your most recent commit, Camille! It's possible some of these comments are already addressed. 

It's looking pretty good! I have a few comments and questions. The text comments are also captured in the video for clarity.

1. When a description is set, I don't see the information icon in table column headers, so there is no way to display on hover.
2. While I do see the information icon in card field headers, but they're different sizes depending on the card theme. The "Form" theme's size looks best (the smallest one).
3. With long field headers, the icon is pushed out of view on cards.
4. The description doesn't appear on hover on the information icon, only on click. Was that a choice for v0?
5. Similarly, the description can only be set in the creator panel, not through a naming panel. Was that a choice for v0? 


https://user-images.githubusercontent.com/69366135/220143202-c2da8a8b-a272-4234-bdf5-0d2f052ba1ef.mp4

**Comment by CamilleLegeron** [comment]
> Thank you, @CamilleLegeron and @berhalak !
> 
> @berhalak pushed the preview to https://grist-column-labels.fly.dev/ so I could play with it. Sorry I didn't get my comments in before your most recent commit, Camille! It's possible some of these comments are already addressed.
> 
> It's looking pretty good! I have a few comments and questions. The text comments are also captured in the video for clarity.
> 
> 1. When a description is set, I don't see the information icon in table column headers, so there is no way to display on hover.
> 2. While I do see the information icon in card field headers, but they're different sizes depending on the card theme. The "Form" theme's size looks best (the smallest one).
> 3. With long field headers, the icon is pushed out of view on cards.
> 4. The description doesn't appear on hover on the information icon, only on click. Was that a choice for v0?
> 5. Similarly, the description can only be set in the creator panel, not through a naming panel. Was that a choice for v0?

Hi Anais,
For the V0 it was decided not to show the description in the table view as there were still uncertain elements, so for your points 1 and 5 it's normal. It will be done in a second time. 
For your points 2 and 3, thank you I had not paid attention to these details, I will change that.
And for your point 4 it the same behavior of the help icon (?)

**Review by berhalak — CHANGES_REQUESTED** [review]

**Review by berhalak — APPROVED** [review]
Thanks @CamilleLegeron

**Review by paulfitz — APPROVED** [review]
Migration looks good.

**Pull request merged**